packages/lisk-transactions/src/transactions/3_vote_transaction.ts

```diff
@@ -0,0 +1,483 @@
+	EntityMap,
+	RequiredState,
+	TransactionResponse,
+} from './base';
+
+const PREFIX_UPVOTE = '+';
+const PREFIX_UNVOTE = '-';
+const MAX_VOTE_PER_ACCOUNT = 101;
+const MIN_VOTE_PER_TX = 1;
+const MAX_VOTE_PER_TX = 33;
+
+export interface VoteAsset {
+	readonly votes: ReadonlyArray<string>;
+}
+
+export interface CreateVoteAssetInput {
+	readonly unvotes?: ReadonlyArray<string>;
+	readonly votes?: ReadonlyArray<string>;
+}
+
+export type CastVoteInput = CreateBaseTransactionInput & CreateVoteAssetInput;
+
+export interface RequiredVoteState extends RequiredState {
+	readonly dependentState?: {
+		readonly [ENTITY_ACCOUNT]: ReadonlyArray<Account>;
+	};
+}
+
+export const voteAssetTypeSchema = {
+	type: 'object',
+	required: ['votes'],
+	properties: {
+		votes: {
+			type: 'array',
+			items: {
+				type: 'string',
+			},
+		},
+	},
+};
+
+export const voteAssetFormatSchema = {
+	type: 'object',
+	required: ['votes'],
+	properties: {
+		votes: {
+			type: 'array',
+			uniqueSignedPublicKeys: true,
+			minItems: MIN_VOTE_PER_TX,
+			maxItems: MAX_VOTE_PER_TX,
+			items: {
+				type: 'string',
+				format: 'signedPublicKey',
+			},
+		},
+	},
+};
+
+const validateInputs = ({
+	votes = [],
+	unvotes = [],
+}: CreateVoteAssetInput): void => {
+	if (!Array.isArray(votes)) {
+		throw new Error(
+			'Please provide a valid votes value. Expected an array if present.',
+		);
+	}
+	if (!Array.isArray(unvotes)) {
+		throw new Error(
+			'Please provide a valid unvotes value. Expected an array if present.',
+		);
+	}
+	validatePublicKeys([...votes, ...unvotes]);
+};
+
+export class VoteTransaction extends BaseTransaction {
+	public readonly containsUniqueData = true;
+	public readonly asset: VoteAsset;
+
+	public constructor(tx: TransactionJSON) {
+		super(tx);
+		const typeValid = validator.validate(voteAssetTypeSchema, tx.asset);
+		const errors = validator.errors
+			? validator.errors.map(
+					error =>
+						new TransactionError(
+							`'${error.dataPath}' ${error.message}`,
+							tx.id,
+							error.dataPath,
+						),
+			  )
+			: [];
+		if (!typeValid) {
+			throw new TransactionMultiError('Invalid field types', tx.id, errors);
+		}
+		this.asset = tx.asset as VoteAsset;
+	}
+
+	public static create(input: CastVoteInput): object {
+		validateInputs(input);
+		const { passphrase, secondPassphrase, votes = [], unvotes = [] } = input;
+
+		const plusPrependedVotes = prependPlusToPublicKeys(votes);
+		const minusPrependedUnvotes = prependMinusToPublicKeys(unvotes);
+		const allVotes: ReadonlyArray<string> = [
+			...plusPrependedVotes,
+			...minusPrependedUnvotes,
+		];
+
+		const transaction = {
+			...createBaseTransaction(input),
+			type: 3,
+			fee: VOTE_FEE.toString(),
+			asset: {
+				votes: allVotes,
+			},
+		};
+
+		if (!passphrase) {
+			return transaction;
+		}
+
+		const {
+			address: senderId,
+			publicKey: senderPublicKey,
+		} = getAddressAndPublicKeyFromPassphrase(passphrase);
+		const transactionWithSenderInfo = {
```

Comment:
Shouldn't this be done in the `createBaseTransaction` function?

Reply:
Is there a reason for why it's done here?

Reply:
one thing it needed to be done specific for vote transaction is  to add the `recipientId` same as `senderId`, but it has to be done after checking the passphrase